### DISCUSSION TITLE
include a local file, when it exist

### DISCUSSION THREAD

**Discussion by FranklinYu** [twpayne/chezmoi#2243] [Show and tell]
From time to time I want to have part of my configuration “local”: optional, not checked into version control, modified only locally. Some of the configuration languages have built-in “include” functionality (e.g. shell, tmux, VimL), but some mini-languages don’t. A notable example is the Git-ignore (such a feature has been asked [here](https://stackoverflow.com/questions/7005142/can-i-include-other-gitignore-file-in-a-gitignore-file-like-include-in-c-li) and [there](https://stackoverflow.com/questions/18750179/gitignore-loads-other-gitignores)).

Specifically, I want to have part of my `~/.config/git/ignore` checked in, part of it remains local. The solution is to generate the `~/.config/git/ignore` from several files (which serve as the source-of-truth). Chezmoi has the helpful template system where I can simply use `dot_config/git/ignore.tmpl`, but I came across a problem: where to put the local file?

The local file can be placed in the source directory, or the target directory. I decided to go with the source directory, so in my `ignore.tmpl` I have

```
{{ if stat (joinPath .chezmoi.sourceDir "local/git-ignore" ) -}}
    {{ include "local/git-ignore" -}}
{{ end }}
```

Notes:

1. `include` requires the file to exist; that’s why I wrapped it in an `if` statement.
2. `stat` function seems to only accept absolute path, unlike `include`, so I have to construct the path with `joinPath`.

If, one wants to put the file in target directory, it’s also possible with Chezmoi. Note that target directory isn’t one of the [template variables](https://www.chezmoi.io/reference/templates/variables/), but we have `.chezmoi.homeDir`, so the template becomes

```
{{ if stat (joinPath .chezmoi.homeDir ".config/git/local-ignore" ) -}}
    {{ include (joinPath .chezmoi.homeDir ".config/git/local-ignore" ) -}}
{{ end }}
```

Your mileage may vary.

**Comment by twpayne** [comment]
Thank you - this is very neat :)
